### PR TITLE
Create Login Controller

### LINKED ISSUE #32: Create Route Login
## Signin Route

Title: POST/signin
labels: Back - Route
Description: Login user account
Protected: No
Required middleWares: No

#### Request body
{
email: ‘y’,
password: ‘z’
}

#### Response
Successs : {message: ‘Logged in Successfully’} & status: 200
Failure: {message: 'invalid email, email doesn’t exist} &  status: 400  // Bad Request
{message: ‘Wrong password’} & status: 401 // validation fails

#### Errors
email not exist? 400
wrong password? 401

#### Schema 
{
email: string()
.email()
.required(),

password: string()
.min(8)
.required()
}

#### Tests
all our status code
200 => success
400 => bad request ‘invalid email’
401 => wrong password

### PR DESCRIPTION
**Pull request by AmranElmasri**
Relates #32 
Close #32

### CODE REVIEW

File: server/src/controllers/user/login.ts [hunk 1]
```diff
@@ -1,5 +1,34 @@
...
+    const { email, password } = await signInSchema.validateAsync(req.body, { abortEarly: false });
+    const { rows, rowCount } = await checkEmailQuery(email);
+    if (rowCount === 0) {
+      throw customError('The email does not exist sign up insted', 401);
+    }
+    const { password: hashedPassword, id, username } = rows[0];
+    const isMatched = compare(password, hashedPassword);
+    if (isMatched) {
+      const token = await signToken({ id, username });
+      res.cookie('token', token, { httpOnly: true, secure: true })
```
Comment (nada-abuzaid) [inline server/src/controllers/user/login.ts]: Edit the secure value to false

File: server/src/controllers/user/login.ts [hunk 2]
```diff
@@ -1,5 +1,34 @@
...
+        .status(200).json({ message: 'login successfully!' });
+    } else {
+      throw customError('please write a correct password', 401);
+    }
+  } catch (err) {
+    if (err.name === 'ValidationError') {
+      const errorList = [];
+      err.details.forEach((error) => errorList.push(error.message));
```
Comment (nada-abuzaid) [inline server/src/controllers/user/login.ts]: Add Error' type to the error parameter

**Review by nada-abuzaid — CHANGES_REQUESTED** [review]

File: server/src/controllers/user/login.ts (hunk 1 above)
Comment (AmranElmasri) [reply server/src/controllers/user/login.ts]: done nada

File: server/src/controllers/user/login.ts (hunk 2 above)
Comment (AmranElmasri) [reply server/src/controllers/user/login.ts]: done nada

**Review by nada-abuzaid — APPROVED** [review]

File: server/src/controllers/index.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import { login } from './user';
+
+// eslint-disable-next-line import/prefer-default-export
```
Comment (Nizar7zak) [inline server/src/controllers/index.ts]: avoid commented lines, instead you can add this rule to your rules in eslint file.

File: server/src/controllers/user/login.ts [hunk 4]
```diff
@@ -1,5 +1,34 @@
...
+import { Request, Response, NextFunction } from 'express';
+import { compare } from 'bcrypt';
+import { checkEmailQuery } from '../../database/ quieres';
+import { signInSchema, signToken } from '../../utils';
+import { customError } from '../errors';
 
-// const login = (req, res) =>{
+const login = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { email, password } = await signInSchema.validateAsync(req.body, { abortEarly: false });
```
Comment (Nizar7zak) [inline server/src/controllers/user/login.ts:9]: you can add option object {abort early} to the schema file, i think it is better than add it here.

File: server/src/controllers/user/login.ts [hunk 5]
```diff
@@ -1,5 +1,34 @@
...
+    const { rows, rowCount } = await checkEmailQuery(email);
+    if (rowCount === 0) {
```
Comment (Nizar7zak) [inline server/src/controllers/user/login.ts]: here you can write a condition like this 
if(!rowCount)

File: server/src/controllers/user/login.ts [hunk 6]
```diff
@@ -1,5 +1,34 @@
...
+      throw customError('The email does not exist sign up insted', 401);
```
Comment (Nizar7zak) [inline server/src/controllers/user/login.ts]: here you should say you have a problem without explain to the client side if this email have an account or not 
` return res.status(400).json({ message: 'Invalid Email or Password' }); `

File: server/src/controllers/user/login.ts [hunk 7]
```diff
@@ -1,5 +1,34 @@
...
+    }
+    const { password: hashedPassword, id, username } = rows[0];
+    const isMatched = compare(password, hashedPassword);
+    if (isMatched) {
+      const token = await signToken({ id, username });
+      res.cookie('token', token, { httpOnly: true, secure: false })
+        .status(200).json({ message: 'login successfully!' });
+    } else {
+      throw customError('please write a correct password', 401);
```
Comment (Nizar7zak) [inline server/src/controllers/user/login.ts]: here you should say you have a problem without explain to the client side if this email have an account or not 
` return res.status(400).json({ message: 'Invalid Email or Password' }); `
 the same thing, moreover your status will be 400 not 401

File: server/src/controllers/user/login.ts [hunk 8]
```diff
@@ -1,5 +1,34 @@
...
+    }
+  } catch (err) {
+    if (err.name === 'ValidationError') {
+      const errorList = [];
+      err.details.forEach((error: Error) => errorList.push(error.message));
```
Comment (Nizar7zak) [inline server/src/controllers/user/login.ts:26]: I suppose if we present the first error come from validation it's better than this way.

**Review by Nizar7zak — CHANGES_REQUESTED** [review]

**Review by karamzomlot — APPROVED** [review]

**Pull request merged**